### ISSUE TITLE
Snapshot in adapter is { adapterOptions: undefined }

### ISSUE THREAD

**Issue by MartinMalinda** [danielspaniel/ember-data-factory-guy#307]
In my adapter method:
```
urlForCreateRecord(modelName, snapshot) {
```

I receive `{adapterOptions: undefined}` in snapshot

It looks like this function passes insufficient data forwards:
https://github.com/danielspaniel/ember-data-factory-guy/blob/f9a10727c48488b3757254a6eca299b194508909/addon/mocks/mock-request.js#L45

I'm doing simple 
```
mockCreate('model');
```

**Comment by danielspaniel** [comment]
Can you write a failing test for that one Martin?   So I can see the way your using it that fails.  If not I can work on it with you if I see more of your code.

**Comment by danielspaniel** [comment]
Got time today .. I can help with test if I can figure out how to recreate what you did

**Comment by MartinMalinda** [comment]
Thank you @danielspaniel. Fortunately I am not blocked on this. Had to move forward pretty quickly so I used plain `mockjax` instead of `mockCreate` as a workaround.

but I'll try creating a reproduction repo real quick

**Comment by MartinMalinda** [comment]
Repo: https://github.com/MartinMalinda/edfg-307

Basic failing test looks like this:
```js
test('urlForCreateRecord works with mockCreate', async function(assert) {
  mockCreate('person');
  const store = this.container.lookup('service:store');

  // const adapter = this.subject();
  // adapter.urlForCreateRecord = function(modelName, snapshot) {
  //   console.log('snapshot is', snapshot);
  //   Ember.assert('record is not available on snapshot', snapshot.record);
  //   return this._super(...arguments);
  // }

  const savedRecord = await Ember.run(() => {
    return store.createRecord('person').save();
  });
  assert.ok(savedRecord);  
});
```

It really fails only when `mockCreate` uses it. Hope it helps!

**Comment by danielspaniel** [comment]
Ok thanks alot .. will give this a try

**Comment by danielspaniel** [comment]
When I do the test I do this 
```javascript 
 return store.createRecord('person').save({adapterOptions: 'dude'});
```
and the adapterOptions are present in the snapshot passed to the adapter method  urlForCreateRecord 
 in snapshot.adapterOptions 

 so I am not sure what you are showing as a failure ? ?  or how you mean that the adapterOptions are not showing up in that method?  me confused

**Comment by MartinMalinda** [comment]
@danielspaniel sorry for the confusion. `adapterOptions` seem to be working fine! It's the rest of the properties on the `snapshot` object that is missing. Probably the title is a bit confusing as I'm not passing, nor needing any `adapterOptions`.

The issue is, that when calling `mockCreate` you only get passed `{ adapterOptions }` as `snapshot` in the `urlForCreateRecord` hook. That is not a `snapshot`. `snapshot` should have a `record` property among other things - which in my case is needed to create a proper url in the adapter.

It works correctly for `save()` but if breaks when `mockCreate` is activated (to handle the mock).

I tried to hint that in the repository with the `Ember.assert`.

**Comment by MartinMalinda** [comment]
There should be a failing test  here: https://github.com/MartinMalinda/edfg-307

**Comment by danielspaniel** [comment]
Currently Martin .. everything works perfectly .. I get the snapshot and it has the record and the adapterOptions  .. 
 I tested on ED 2.13 and 2.15  .. not sure if that matters or not though 
 but not sure what else I can do ??

here is my test 

```javascript 
test('snapshot has record and adapterOptions in adapter#urlForCreateRecord', async function(assert) {
  mockCreate('user');

  let adapter        = FactoryGuy.store.adapterFor('user'),
      adapterOptions = { name: 'Bob' };

  adapter.urlForCreateRecord = function(modelName, snapshot) {
    assert.ok(snapshot.record instanceof Model);
    assert.deepEqual(snapshot.adapterOptions, adapterOptions);
    return '/users';
  };

  return await Ember.run(() => {
    return FactoryGuy.store.createRecord('user').save({ adapterOptions });
  });
});
```

**Comment by MartinMalinda** [comment]
@danielspaniel thanks. I have updated my the unit test in my repo based on this and still fails. It passes at first for `save()` and fails shortly after for `mockCreate()`.

I'm on EDFG 2.13.8 and ED 2.16 (tested also on 2.14)
https://github.com/MartinMalinda/edfg-307/blob/master/package.json#L33

Thanks a lot for looking into this. As I've said before I do have a workaround, it's not a blocker for me! Let's see if someone else experiences this. Maybe I'll try to add a PR with failing test later. So far the failing scenario is available in the repo.

**Comment by danielspaniel** [comment]
You were right Martin, this is a bug. Your repo helped me figure out the problem. 
 I have done a bunch of things to make this work.   Still not done, and the answer will not be perfect but as close as I can get to it.

**Comment by danielspaniel** [comment]
This issue is actually never going to be fixed 100%, because of 1 major issue that i can not control. 

The way adapter method urlForCreateRecord works is that it returns a url given a snapshot and adapterOptions and other stuff.   
FactoryGuy needs a url to setup pretender. 

How can FactoryGuy get its hands on the model if you use this style of mockCreate:

 mockCreate('user') 

The answer is "it can't".  Since the mock has no model available ( and no way to find it since there is no id ) in the store. 

One answer is to make that model available like this:

```javascript 
let user = makeNew('user')
mockCreate(user);  // you can now pass a real model to mockCreate 
```

Now that mockCreate knows the model, it can pass it along to the urlForCreateRecord method in the adapter. 

 So for this to work for you you will have to do this: 

```javascript 
let user = makeNew('user')
mockCreate(user).withAdapterOptions({blah: 1});
```

and now both the model and adapterOptions will be used found in the snapshot

But, what if you can't make the new record first and pass it into the mockCreate method? 

in that case your will have to use sinon to mock the return from that urlForCreateRecord like this: 

```javascript
import sinon from 'sinon'
import FactoryGuy, { mockCreate } from 'ember-data-factory-guy';

test('create user as zombie', function() {
  let adapter = FactoryGuy.store.adapterFor('user');
  sinon.stub(adapter, 'urlForCreateRecord').returns('/createUserAsZombiePlease');
  mockCreate('user'); // don't have the user record .. just model name
  ...
  adapter.urlForCreateRecord.restore();
});
```

In other words, you are giving up on the snapshot being perfect and just avoiding the urlForCreateRecord method by telling it what to return.  It's the only option I can think of. But it is not so terrible because you can test the urlForCreateMethod separately in a unit test to be sure it returns the correct url under different circumstances, so you don't really need ( need ) to test that url  in other tests. Though usually factory guy can just do the right thing and save you the trouble of the sinon.stubbing .. this ONE particuar case I can not save you the hassle. 


Can you confirm you understand what I am saying and/or close this or keep asking questions?

**Comment by danielspaniel** [comment]
Going to close this .. because I think it is as fixed as I can get it, but feel free to reopen if it is still a problem.
